### PR TITLE
Issue #8690: Specify the frecency threshold for the fetched top frecent sites

### LINKED ISSUE #8690: Allow TopSitesConfig to specify the frecency threshold
In https://github.com/mozilla-mobile/fenix/issues/14302, we want to align with how desktop is displaying the top visited sites as top sites. 

In Desktop, this is the frencey threshold used to determine if a visited site should be displayed as a top site. https://searchfox.org/mozilla-central/rev/919607a3610222099fbfb0113c98b77888ebcbfb/browser/components/newtab/lib/TopSitesFeed.jsm#79

In https://github.com/mozilla/application-services/issues/3635, we're including the frecency score as part of the TopFrecentSiteInfo. 

In AC, we want to add an option to specify the frecency threshold that can be specified by the consumer when getting the list of top sites with top frecent sites.

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #8690. Requires AS bump with https://github.com/mozilla/application-services/pull/3645. 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt [hunk 1]
```diff
@@ -67,7 +68,8 @@ class DefaultTopSitesStorage(
 
     override suspend fun getTopSites(
         totalSites: Int,
-        includeFrecent: Boolean
+        includeFrecent: Boolean,
+        frecencyThreshold: FrecencyThresholdOption
```
Comment (jonalmeida) [inline components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt:72]: Does this option change? Maybe just put it in the constructor or leave it internal as part of our implicit feature expectations.

Comment (gabrielluong) [reply components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt:72]: It doesn't change, but we also fetch the option from the TopSiteConfig

File: buildSrc/src/main/java/Dependencies.kt [hunk 2]
```diff
@@ -30,7 +30,7 @@ object Versions {
     const val disklrucache = "2.0.2"
     const val leakcanary = "2.4"
 
-    const val mozilla_appservices = "63.0.0"
```
Comment (grigoryk) [inline buildSrc/src/main/java/Dependencies.kt:33]: Can we do this bump in a separate PR, to make sure nothing breaks? Or do a round of testing here, that works as well.

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt [hunk 3]
```diff
@@ -31,6 +32,14 @@ internal fun mozilla.components.concept.sync.SyncAuthInfo.into(): SyncAuthInfo {
     )
 }
 
+/**
+ * Conversion from a generic [FrecencyThresholdOption] into its richer comrade within the 'places' lib.
```
Comment (grigoryk) [inline components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt:36]: Not sure about the `richer` part, but I guess it could be! :) Hopefully they're maintained in lockstep though.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 4]
```diff
@@ -69,13 +69,18 @@ interface HistoryStorage : Storage {
     ): List<VisitInfo>
 
     /**
-     * Returns a list of the top frecent site infos limited by the given number of items
-     * sorted by most to least frecent.
+     * Returns a list of the top frecent site infos limited by the given number of items and
+     * frecency threshold sorted by most to least frecent.
      *
      * @param numItems the number of top frecent sites to return in the list.
+     * @param frecencyThreshold frecency threshold options for filtering visited sites based on
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:76]: s/options/option

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 5]
```diff
@@ -165,6 +172,18 @@ data class TopFrecentSiteInfo(
     val title: String?
 )
 
+/**
+ * Frecency threshold options for fetching top frecent sites. Requests a page that was visited
+ * with a frecency score greater or equal to the [value].
+ */
+enum class FrecencyThresholdOption(val value: Long) {
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:179]: Do you anticipate this growing over time? It could, in theory... but if not, a simple flag would have been simpler.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt [hunk 6]
```diff
@@ -165,6 +172,18 @@ data class TopFrecentSiteInfo(
...
+    /** Returns all visited pages. */
+    NONE(0),
+
+    /** Skip visited pages that were only visited once. */
+    SKIP_ONE_TIME_PAGES(101)
```
Comment (grigoryk) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:184]: I don't think you need the actual frecency scores here? These are already encapsulated in the A-S counterpart.

File: components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesConfig.kt [hunk 7]
```diff
@@ -4,14 +4,20 @@
...
  * @property totalSites A total number of sites that will be displayed.
  * @property includeFrecent If true, includes frecent top site results.
+ * @property frecencyThreshold If [includeFrecent] is true, only visited sites with a frecency
+ * score above the given threshold will be returned. Defaults to 0 to return all visited pages
+ * sorted by most to least frecent.
  */
 data class TopSitesConfig(
     val totalSites: Int,
-    val includeFrecent: Boolean
+    val includeFrecent: Boolean,
```
Comment (grigoryk) [inline components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesConfig.kt:21]: You should simplify this down to a single optional param, something like `frecencyConfig: FrecencyThresholdOption?`. If it's absent, that would be equivalent to `includeFrecent = false`.

Otherwise, you could have an illegal set of values passed into this data class, e.g. `includeFrecent=false, frecencyThreshold=SKIP_ONE_TIME_PAGES`.

This is the pattern we already follow elsewhere (e.g. fxa).

File: components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesStorage.kt [hunk 8]
```diff
@@ -41,8 +42,14 @@ interface TopSitesStorage : Observable<TopSitesStorage.Observer> {
      *
      * @param totalSites A total number of sites that will be retrieve if possible.
      * @param includeFrecent If true, includes frecent top site results.
+     * @param frecencyThreshold If [includeFrecent] is true, frecency threshold options for
+     * filtering visited sites based on their frecency score.
      */
-    suspend fun getTopSites(totalSites: Int, includeFrecent: Boolean): List<TopSite>
+    suspend fun getTopSites(
+        totalSites: Int,
+        includeFrecent: Boolean,
```
Comment (grigoryk) [inline components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesStorage.kt:50]: Same comment re simplifying this down to a single `frecencyConfig: FrecencyThresholdOption?` value.

File: docs/changelog.md [hunk 9]
```diff
@@ -46,6 +46,7 @@ permalink: /changelog/
 
 * **feature-top-sites**
   * Added `RenameTopSiteUseCase` to rename pinned site entries. [#8751](https://github.com/mozilla-mobile/android-components/issues/8751)
+  * ⚠️ **This is a breaking change**: Adds a new parameter `frecencyThreshold` to `TopSitesConfig` and `TopSitesStorage.getTopSites` to specify the frecency threshold for the returned list of top frecent sites see [#8690](https://github.com/mozilla-mobile/android-components/issues/8690).
```
Comment (grigoryk) [inline docs/changelog.md:49]: You could give these default values to avoid breaking consumers.

**Review by grigoryk — CHANGES_REQUESTED** [review]
Overall this looks good! I think we should simplify the APIs a bit to avoid having illegal combinations around, and make sure to separately test the a-s bump (I'd prefer that in a separate PR, actually).

File: buildSrc/src/main/java/Dependencies.kt (hunk 2 above)
Comment (gabrielluong) [reply buildSrc/src/main/java/Dependencies.kt:33]: Let's do that here since it's a breaking change from AS.

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt (hunk 3 above)
Comment (gabrielluong) [reply components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt:36]: Gonna keep it as-is since it's consistent with the VisitType `into()`.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt (hunk 6 above)
Comment (gabrielluong) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:184]: Removed the value.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt (hunk 5 above)
Comment (gabrielluong) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/HistoryStorage.kt:179]: Only in theory. I am accounting for potential changes, but don't currently anticipate anything.

Realistically, this could've been just a value (101) that we feed into the AS API or hardcoded somewhere, but I didn't want to expose a magical number like 101 which might not signify anything to consumers.

File: components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesConfig.kt (hunk 7 above)
Comment (gabrielluong) [reply components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesConfig.kt:21]: Merge the 2 params into `frecencyConfig: FrecencyThresholdOption?`

File: components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesStorage.kt (hunk 8 above)
Comment (gabrielluong) [reply components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesStorage.kt:50]: Done

File: buildSrc/src/main/java/Dependencies.kt (hunk 2 above)
Comment (grigoryk) [reply buildSrc/src/main/java/Dependencies.kt:33]: Sure thing. Let's just make sure we do a full smoke test of A-S-depended functionality here - logins, fxa, sync, send tab, etc.

**Review by grigoryk — APPROVED** [review]
Looks good! Just some nits and random musings.

🚢🚢🚢

**Comment by grigoryk** [comment]
@gabrielluong i'll do a round of testing for the a-s version bump before we land this.

File: components/browser/storage-memory/src/main/java/mozilla/components/browser/storage/memory/InMemoryHistoryStorage.kt [hunk 10]
```diff
@@ -80,57 +87,78 @@ class InMemoryHistoryStorage : HistoryStorage {
...
+                } else {
+                    matchedUrls[it.url] = it.score
+                }
             }
+            // Calculate maxScore so that we can invert our scoring.
+            // Lower Levenshtein distance should produce a higher score.
+            val maxScore =
+                urlMatches.maxByOrNull { it.score }?.score ?: return@synchronized listOf()
+
+            // TODO exclude non-matching results entirely? Score that implies complete mismatch.
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply components/browser/storage-memory/src/main/java/mozilla/components/browser/storage/memory/InMemoryHistoryStorage.kt:139]: I think this was an unintended change from the auto-formatting

File: components/browser/storage-memory/src/main/java/mozilla/components/browser/storage/memory/InMemoryHistoryStorage.kt [hunk 11]
```diff
@@ -80,57 +87,78 @@ class InMemoryHistoryStorage : HistoryStorage {
...
+    override suspend fun getTopFrecentSites(
+        numItems: Int,
+        frecencyThreshold: FrecencyThresholdOption
+    ): List<TopFrecentSiteInfo> {
         throw UnsupportedOperationException("getTopFrecentSites is not yet supported by the in-memory history storage")
     }
 
-    override fun getSuggestions(query: String, limit: Int): List<SearchResult> = synchronized(pages + pageMeta) {
-        data class Hit(val url: String, val score: Int)
+    override fun getSuggestions(query: String, limit: Int): List<SearchResult> =
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply components/browser/storage-memory/src/main/java/mozilla/components/browser/storage/memory/InMemoryHistoryStorage.kt:116]: Fixed.

File: components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt [hunk 12]
```diff
@@ -301,7 +303,7 @@ class DefaultTopSitesStorageTest {
             )
         )
 
-        val topSites = defaultTopSitesStorage.getTopSites(5, true)
+        val topSites = defaultTopSitesStorage.getTopSites(5, frecencyConfig = FrecencyThresholdOption.NONE)
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt:306]: Added

`verify(historyStorage).getTopFrecentSites(5, frecencyThreshold = FrecencyThresholdOption.NONE)`

File: components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt [hunk 13]
```diff
@@ -156,29 +157,48 @@ class PlacesHistoryStorageTest {
...
+
+        infos = history.getTopFrecentSites(100, frecencyThreshold = FrecencyThresholdOption.NONE)
         assertEquals(5, infos.size)
         assertEquals("https://www.mozilla.com/foo/bar/baz", infos[0].url)
         assertEquals("https://www.example.com/123", infos[1].url)
         assertEquals("https://news.ycombinator.com/", infos[2].url)
         assertEquals("https://mozilla.com/a1/b2/c3", infos[3].url)
         assertEquals("https://www.example.com/12345", infos[4].url)
+
+        infos = history.getTopFrecentSites(100, frecencyThreshold = FrecencyThresholdOption.SKIP_ONE_TIME_PAGES)
```
Comment (grigoryk) [inline components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt:198]: Will it work with negative values? e.g. `history.getTopFrecentSites(-4, frecencyThreshold = FrecencyThresholdOption.NONE)`

Textbook unit test :)

File: components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt [hunk 14]
```diff
@@ -213,15 +215,15 @@ class DefaultTopSitesStorageTest {
 
         val frecentSite2 = TopFrecentSiteInfo("https://example.com", "Example")
         val frecentSite3 = TopFrecentSiteInfo("https://getpocket.com", "Pocket")
-        whenever(historyStorage.getTopFrecentSites(anyInt())).thenReturn(
+        whenever(historyStorage.getTopFrecentSites(anyInt(), any())).thenReturn(
```
Comment (grigoryk) [inline components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt:218]: I really recommend doing away with the mocks, and just using the real storage implementation instead. It's a little bit more working setting things up, but really not much more - you'd need to actually insert history items and visits yourself to prep the storage.

For an example of how to set this up, see `browser-storage-sync` component. Specifically, you may need to copy-paste some bits from `build.gradle`:
```
// These dependencies are part of this module's public API.
    api(Dependencies.mozilla_places) {
        // Use our own version of the Glean dependency,
        // which might be different from the version declared by A-S.
        exclude group: 'org.mozilla.components', module: 'service-glean'
    }
...
testImplementation Dependencies.mozilla_places
testImplementation Dependencies.mozilla_full_megazord_forUnitTests
testImplementation Dependencies.mozilla_glean_forUnitTests
...
```

(it may work without the glean stuff, try that first; i don't recall the details right now)

The result should be a much more robust and useful test suite which exercises our whole stack! This will become a good integration test, instead of a "we're calling the right APIs" unit test. The latter is helpful, but the former actually gives you confidence that the system is working correctly.

File: components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt (hunk 13 above)
Comment (gabrielluong) [reply components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt:198]: I added 

```
if (numItems <= 0) {
   return emptyList()
}
```

but I am wondering if we would prefer doing a `require(numItems >= 0)`

File: components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt (hunk 14 above)
Comment (gabrielluong) [reply components/feature/top-sites/src/test/java/mozilla/components/feature/top/sites/DefaultTopSitesStorageTest.kt:262]: I spent some time on this, but ran into some issues that I have not managed to figure out yet. I will spin this task into a new issue https://github.com/mozilla-mobile/android-components/issues/9501 to not block this any further.
